HULKs/hulk
# Cihaviorsimulator

## Introduced Changes

This PR adds tools to the lua API with which behavior scenarios can implement pass/fail conditions.
These conditions can then be checked in a CI job to prevent behavior regressions.

The behavior simulator's CLI interface was changed, it now takes a command: `run` or `serve`, both of which still take the path to a scenario file.
`run` exits after the simulation finishes, `serve` replicates the old behavior.

Fixes #551

## ToDo / Known Issues

- [x] Add way to run simulator without starting a blocking communication server
- [x] Allow generating errors from lua
- [x] Run scenarios in CI
  - [x] Get inspect.lua into CI environment
  - [x] Figure out how to organize running multiple scenarios
- [x] Don't terminate program on error in `serve` mode

## Ideas for Next Iterations (Not This PR)



## How to Test

Run scenario like so:
```
cargo run --manifest-path tools/behavior_simulator/Cargo.toml --release run tests/behavior/golden_goal.lua
```
This should run without errors and exit with code 0.

Now break it in some way, for example by either never setting the goal scored flag in the scenario file or never assigning a striker.

The above command should then show an error message and exit with code 1.
